lasagne/layers/transform.py

```diff
@@ -0,0 +1,200 @@
+    """
+    Spatial transformer layer [1]_.
+
+    Parameters
+    ----------
+    input : a :class:`Layer` instance
+        The input where the affine transformation is applied. This should
+        have convolution format, i.e (num_batch, channels, height, width).
+    localization_network: : a :class:`Layer` instance
+        The network that calculates the parameters of the affine
+        transformation. See the example for how to initialize to the identity
+        transform. Note that all parameters are clipped to lie in the range
+        ``[-1, 1]``, similar to [1]_.
+    downsample_factor : float
+        Determines the size of the output image. A value of 1 will keep the
+        original size of the input. Values larger than 1 will down sample the
+        input. Values below 1 will up sample the input.
+
+    References
+    ----------
+    .. [1]  Spatial Transformer Networks
+            Max Jaderberg, Karen Simonyan, Andrew Zisserman, Koray Kavukcuoglu
+            Submitted on 5 Jun 2015
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> import lasagne
+    >>> b = np.zeros((2, 3), dtype='float32')
+    >>> b[0, 0] = 1
+    >>> b[1, 1] = 1
+    >>> b = b.flatten()  # identity transform
+    >>> W = lasagne.init.Constant(0.0)
+    >>> l_in = lasagne.layers.InputLayer((None, 3, 28, 28))
+    >>> l_loc = lasagne.layers.DenseLayer(l_in, num_units=6, W=W, b=b,
+    ... nonlinearity=None)
+    >>> l_trans = lasagne.layers.TransformerLayer(l_in, l_loc)
+    """
+    def __init__(
+            self, input, localization_network, downsample_factor=1, **kwargs):
+        super(TransformerLayer, self).__init__(
+            [input, localization_network], **kwargs)
+        self.downsample_factor = downsample_factor
+
+        input_shp, loc_shp = self.input_shapes
+
+        if loc_shp[-1] != 6 or len(loc_shp) != 2:
+            raise ValueError("The localization network shoud have "
+                             "outputformat [num_batch, 6]")
+        if len(input_shp) != 4:
+            raise ValueError('The input network must have 4 dimensions: '
+                             '[num_batch, channels, height width]')
+
+    def get_output_shape_for(self, input_shapes):
+        shp = input_shapes[0]
+        dsf = self.downsample_factor
+        return tuple(list(shp[:2]) + [int(s//dsf) for s in shp[2:]])
+
+    def get_output_for(self, inputs, deterministic=False, **kwargs):
+        # theta should be shape (num_batch, 2, 3)
+        # see eq. (1) and sec 3.1 in [1]
+        input, theta = inputs
+        return _transform(theta, input, self.downsample_factor)
+
+
+def _transform(theta, input, downsample_factor):
+    num_batch, num_channels, height, width = input.shape
+    theta = T.reshape(theta, (-1, 2, 3))
+
+    height_f = T.cast(height, 'float32')
+    width_f = T.cast(width, 'float32')
+
+    out_height = T.cast(height_f // downsample_factor, 'int64')
+    out_width = T.cast(width_f // downsample_factor, 'int64')
```

Comment:
> Will it still be copied to cpu if i use int32?

Not sure. But I'm also not so sure if it is helpful to have that shape computation on GPU.

If you don't need integers in the end, you could also just cast the symbolic shapes and downsample factor to `floatX`:
```python
height = T.cast(height, theano.config.floatX)
width = T.cast(width, theano.config.floatX)
downsample_factor = np.asarray(downsample_factor, dtype=theano.config.floatX)
```